### ISSUE TITLE
migrate fails with  j.literal is not a function

### ISSUE THREAD

**Issue by zowers** [webpack/webpack-cli#145]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
migrate fails with
```
 √ Reading webpack config
 > Migrating config from v1 to v2
   → j.literal is not a function
   × loadersTransform
     → j.literal is not a function
     resolveTransform
     removeJsonLoaderTransform
     uglifyJsPluginTransform
     loaderOptionsPluginTransform
     bannerPluginTransform
     extractTextPluginTransform
     removeDeprecatedPluginsTransform
✖ ︎Migration aborted due to some errors
{ TypeError: j.literal is not a function
    at Object.createLiteral (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\lib\transformations\utils.js:111:11)
    at query.split.map.option (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\lib\transformations\loaders\loaders.js:25:53)
    at Array.map (native)
    at NodePath.createLoaderWithQuery (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\lib\transformations\loaders\loaders.js:21:34)
    at NodePath.<anonymous> (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\node_modules\jscodeshift\dist\collections\Node.js:144:47)
    at __paths.forEach (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\node_modules\jscodeshift\dist\Collection.js:76:36)
    at Array.forEach (native)
    at Collection.forEach (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\node_modules\jscodeshift\dist\Collection.js:75:18)
    at Collection.replaceWith (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\node_modules\jscodeshift\dist\collections\Node.js:142:17)
    at Collection.typedMethod (E:\Users\user\AppData\Roaming\npm\node_modules\webpack-cli\node_modules\jscodeshift\dist\Collection.js:376:43)
  context:
   { source: 'const argv = require( "yargs" ).argv;\r\nconst webpack = require( "webpack" );\r\nconst cssnano = require( "cssnano" );\r\nconst HtmlWebpackPlugin = require( "html-webpack-plugin" );\r\nconst ExtractTextPlugin = require( "extract-text-webpack-plugin" );\r\nconst project = require( "./project.config" );\r\nconst debug = require( "debug" )( "app:config:webpack" );\r\n\r\nconst __DEV__ = project.globals.__DEV__;\r\nconst __PROD__ = project.globals.__PROD__;\r\nconst __TEST__ = project.globals.__TEST__;\r\n\r\ndebug( "Creating configuration." );\r\nconst webpackConfig = {\r\n  name: "client",\r\n  target: "web",\r\n  devtool: project.compiler_devtool,\r\n  resolve: {\r\n    root: project.paths.client(),\r\n    extensions: [ "", ".js", ".jsx", ".json" ]\r\n  },\r\n  module: {}\r\n};\r\n// ------------------------------------\r\n// Entry Points\r\n// ------------------------------------\r\nconst APP_ENTRY = project.paths.client( "main.js" );\r\n\r\nwebpackConfig.entry = {\r\n  app: __DEV__\r\n    ? [ APP_ENTRY ].concat( `webpack-hot-middleware/client?path=${project.compiler_public_path}__webpack_hmr` )\r\n    : [ APP_ENTRY ],\r\n  vendor: project.compiler_vendors\r\n};\r\n\r\n// ------------------------------------\r\n// Bundle Output\r\n// ------------------------------------\r\nwebpackConfig.output = {\r\n  filename: `[name].[${project.compiler_hash_type}].js`,\r\n  path: project.paths.dist(),\r\n  publicPath: project.compiler_public_path\r\n};\r\n\r\n// ------------------------------------\r\n// Externals\r\n// ------------------------------------\r\nwebpackConfig.externals = {};\r\nwebpackConfig.externals["react/lib/ExecutionEnvironment"] = true;\r\nwebpackConfig.externals["react/lib/ReactContext"] = true;\r\nwebpackConfig.externals["react/addons"] = true;\r\n\r\n// ------------------------------------\r\n// Plugins\r\n// ------------------------------------\r\nwebpackConfig.plugins = [\r\n  new webpack.DefinePlugin( project.globals ),\r\n  new HtmlWebpackPlugin( {\r\n    template: project.paths.client( "index.html" ),\r\n    hash: false,\r\n    favicon: project.paths.public( "favicon.ico" ),\r\n    filename: "index.html",\r\n    inject: "body",\r\n    minify: {\r\n      collapseWhitespace: true\r\n    }\r\n  } )\r\n];\r\n\r\n// Ensure that the compiler exits on errors during testing so that\r\n// they do not get skipped and misreported.\r\nif( __TEST__ && !argv.watch ) {\r\n  webpackConfig.plugins.push( function () {\r\n    this.plugin( "done", function ( stats ) {\r\n      if( stats.compilation.errors.length ) {\r\n        // Pretend no assets were generated. This prevents the tests\r\n        // from running making it clear that there were warnings.\r\n        throw new Error(\r\n          stats.compilation.errors.map( err => err.message || err )\r\n        );\r\n      }\r\n    } );\r\n  } );\r\n}\r\n\r\nif( __DEV__ ) {\r\n  debug( "Enabling plugins for live development (HMR, NoErrors)." );\r\n  webpackConfig.plugins.push(\r\n    new webpack.HotModuleReplacementPlugin(),\r\n    new webpack.NoErrorsPlugin()\r\n  );\r\n} else if( __PROD__ ) {\r\n  debug( "Enabling plugins for production (OccurenceOrder, Dedupe & UglifyJS)." );\r\n  webpackConfig.plugins.push(\r\n    new webpack.optimize.OccurrenceOrderPlugin(),\r\n    new webpack.optimize.DedupePlugin(),\r\n    new webpack.optimize.UglifyJsPlugin( {\r\n        mangle: false,\r\n        compress: {\r\n            unused: false, // do not remove: responsible for named functions\r\n            dead_code: true,\r\n            sequences: false,\r\n            drop_console: true,\r\n            drop_debugger: false,\r\n            hoist_funs: true,\r\n            join_vars: false,\r\n            side_effects: false,\r\n            warnings: false,\r\n            global_defs: {\r\n                DEBUG: false\r\n            }\r\n        },\r\n        output: {\r\n            beautify: false,\r\n            semicolons: true,\r\n            comments: false,\r\n            indent_level: 0,\r\n            ascii_only: true,\r\n            width: 80,\r\n            max_line_len: 80\r\n            //preserve_line: true,\r\n        }\r\n    } )\r\n  );\r\n}\r\n\r\n// Don\'t split bundles during testing, since we only want import one bundle\r\nif( !__TEST__ ) {\r\n  webpackConfig.plugins.push(\r\n    new webpack.optimize.CommonsChunkPlugin( {\r\n      names: [ "vendor" ]\r\n    } )\r\n  );\r\n}\r\n\r\n// ------------------------------------\r\n// Loaders\r\n// ------------------------------------\r\n// JavaScript / JSON\r\nwebpackConfig.module.loaders = [ {\r\n  test: /\\.(js|jsx)$/,\r\n  exclude: /node_modules/,\r\n  loader: "babel",\r\n  query: project.compiler_babel\r\n}, {\r\n  test: /\\.json$/,\r\n  loader: "json"\r\n} ];\r\n\r\n// ------------------------------------\r\n// Style Loaders\r\n// ------------------------------------\r\n// We use cssnano with the postcss loader, so we tell\r\n// css-loader not to duplicate minimization.\r\nconst BASE_CSS_LOADER =  "css?importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]&sourceMap&-minimize";\r\n//\r\n\r\nwebpackConfig.module.loaders.push( {\r\n  test: /\\.scss$/,\r\n  exclude: null,\r\n  loaders: [\r\n    "style-loader",\r\n    BASE_CSS_LOADER,\r\n    "resolve-url-loader",\r\n    "postcss",\r\n    "sass?sourceMap"\r\n  ]\r\n} );\r\nwebpackConfig.module.loaders.push( {\r\n  test: /\\.css$/,\r\n  exclude: null,\r\n  loaders: [\r\n    "style-loader",\r\n    BASE_CSS_LOADER,\r\n    "resolve-url-loader",\r\n    "postcss"\r\n  ]\r\n} );\r\n\r\nwebpackConfig.resolveUrlLoader =  {\r\n  root: "./src",\r\n};\r\n\r\nwebpackConfig.sassLoader = {\r\n  includePaths: project.paths.client( "styles" )\r\n};\r\n\r\nwebpackConfig.postcss = [\r\n  cssnano( {\r\n    autoprefixer: {\r\n      add: true,\r\n      remove: true,\r\n      browsers: [ "last 2 versions" ]\r\n    },\r\n    discardComments: {\r\n      removeAll: true\r\n    },\r\n    discardUnused: false,\r\n    mergeIdents: false,\r\n    reduceIdents: false,\r\n    safe: true,\r\n    sourcemap: true\r\n  } )\r\n];\r\n\r\n// File loaders\r\n/* eslint-disable */\r\nwebpackConfig.module.loaders.push(\r\n  { test: /\\.woff(\\?.*)?$/,  loader: \'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff\' },\r\n  { test: /\\.woff2(\\?.*)?$/, loader: \'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff2\' },\r\n  { test: /\\.otf(\\?.*)?$/,   loader: \'file?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=font/opentype\' },\r\n  { test: /\\.ttf(\\?.*)?$/,   loader: \'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/octet-stream\' },\r\n  { test: /\\.eot(\\?.*)?$/,   loader: \'file?prefix=fonts/&name=[path][name].[ext]\' },\r\n  { test: /\\.svg(\\?.*)?$/,   loader: \'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=image/svg+xml\' },\r\n  { test: /\\.(png|jpg)$/,    loader: \'url?limit=8192\' }\r\n)\r\n/* eslint-enable */\r\n\r\n// ------------------------------------\r\n// Finalize Configuration\r\n// ------------------------------------\r\n// when we don\'t know the public path (we know it only when HMR is enabled [in development]) we\r\n// need to use the extractTextPlugin to fix this issue:\r\n// http://stackoverflow.com/questions/34133808/webpack-ots-parsing-error-loading-fonts/34133809#34133809\r\nif( !__DEV__ ) {\r\n  debug( "Applying ExtractTextPlugin to CSS loaders." );\r\n  webpackConfig.module.loaders.filter( ( loader ) =>\r\n    loader.loaders && loader.loaders.find( ( name ) => /css/.test( name.split( "?" )[0] ) )\r\n  ).forEach( ( loader ) => {\r\n    const first = loader.loaders[0];\r\n    const rest = loader.loaders.slice( 1 );\r\n    loader.loader = ExtractTextPlugin.extract( first, rest.join( "!" ) );\r\n    delete loader.loaders;\r\n  } );\r\n\r\n  webpackConfig.plugins.push(\r\n    new ExtractTextPlugin( "[name].[contenthash].css", {\r\n      allChunks: true\r\n    } )\r\n  );\r\n}\r\n\r\nmodule.exports = webpackConfig;\r\n',
```

**If the current behavior is a bug, please provide the steps to reproduce.**
run `webpack-cli migrate ./webpack.config.js` on the file:
```js
const argv = require( "yargs" ).argv;
const webpack = require( "webpack" );
const cssnano = require( "cssnano" );
const HtmlWebpackPlugin = require( "html-webpack-plugin" );
const ExtractTextPlugin = require( "extract-text-webpack-plugin" );
const project = require( "./project.config" );
const debug = require( "debug" )( "app:config:webpack" );

const __DEV__ = project.globals.__DEV__;
const __PROD__ = project.globals.__PROD__;
const __TEST__ = project.globals.__TEST__;

debug( "Creating configuration." );
const webpackConfig = {
  name: "client",
  target: "web",
  devtool: project.compiler_devtool,
  resolve: {
    root: project.paths.client(),
    extensions: [ "", ".js", ".jsx", ".json" ]
  },
  module: {}
};
// ------------------------------------
// Entry Points
// ------------------------------------
const APP_ENTRY = project.paths.client( "main.js" );

webpackConfig.entry = {
  app: __DEV__
    ? [ APP_ENTRY ].concat( `webpack-hot-middleware/client?path=${project.compiler_public_path}__webpack_hmr` )
    : [ APP_ENTRY ],
  vendor: project.compiler_vendors
};

// ------------------------------------
// Bundle Output
// ------------------------------------
webpackConfig.output = {
  filename: `[name].[${project.compiler_hash_type}].js`,
  path: project.paths.dist(),
  publicPath: project.compiler_public_path
};

// ------------------------------------
// Externals
// ------------------------------------
webpackConfig.externals = {};
webpackConfig.externals["react/lib/ExecutionEnvironment"] = true;
webpackConfig.externals["react/lib/ReactContext"] = true;
webpackConfig.externals["react/addons"] = true;

// ------------------------------------
// Plugins
// ------------------------------------
webpackConfig.plugins = [
  new webpack.DefinePlugin( project.globals ),
  new HtmlWebpackPlugin( {
    template: project.paths.client( "index.html" ),
    hash: false,
    favicon: project.paths.public( "favicon.ico" ),
    filename: "index.html",
    inject: "body",
    minify: {
      collapseWhitespace: true
    }
  } )
];

// Ensure that the compiler exits on errors during testing so that
// they do not get skipped and misreported.
if( __TEST__ && !argv.watch ) {
  webpackConfig.plugins.push( function () {
    this.plugin( "done", function ( stats ) {
      if( stats.compilation.errors.length ) {
        // Pretend no assets were generated. This prevents the tests
        // from running making it clear that there were warnings.
        throw new Error(
          stats.compilation.errors.map( err => err.message || err )
        );
      }
    } );
  } );
}

if( __DEV__ ) {
  debug( "Enabling plugins for live development (HMR, NoErrors)." );
  webpackConfig.plugins.push(
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NoErrorsPlugin()
  );
} else if( __PROD__ ) {
  debug( "Enabling plugins for production (OccurenceOrder, Dedupe & UglifyJS)." );
  webpackConfig.plugins.push(
    new webpack.optimize.OccurrenceOrderPlugin(),
    new webpack.optimize.DedupePlugin(),
    new webpack.optimize.UglifyJsPlugin( {
        mangle: false,
        compress: {
            unused: false, // do not remove: responsible for named functions
            dead_code: true,
            sequences: false,
            drop_console: true,
            drop_debugger: false,
            hoist_funs: true,
            join_vars: false,
            side_effects: false,
            warnings: false,
            global_defs: {
                DEBUG: false
            }
        },
        output: {
            beautify: false,
            semicolons: true,
            comments: false,
            indent_level: 0,
            ascii_only: true,
            width: 80,
            max_line_len: 80
            //preserve_line: true,
        }
    } )
  );
}

// Don't split bundles during testing, since we only want import one bundle
if( !__TEST__ ) {
  webpackConfig.plugins.push(
    new webpack.optimize.CommonsChunkPlugin( {
      names: [ "vendor" ]
    } )
  );
}

// ------------------------------------
// Loaders
// ------------------------------------
// JavaScript / JSON
webpackConfig.module.loaders = [ {
  test: /\.(js|jsx)$/,
  exclude: /node_modules/,
  loader: "babel",
  query: project.compiler_babel
}, {
  test: /\.json$/,
  loader: "json"
} ];

// ------------------------------------
// Style Loaders
// ------------------------------------
// We use cssnano with the postcss loader, so we tell
// css-loader not to duplicate minimization.
const BASE_CSS_LOADER =  "css?importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]&sourceMap&-minimize";
//

webpackConfig.module.loaders.push( {
  test: /\.scss$/,
  exclude: null,
  loaders: [
    "style-loader",
    BASE_CSS_LOADER,
    "resolve-url-loader",
    "postcss",
    "sass?sourceMap"
  ]
} );
webpackConfig.module.loaders.push( {
  test: /\.css$/,
  exclude: null,
  loaders: [
    "style-loader",
    BASE_CSS_LOADER,
    "resolve-url-loader",
    "postcss"
  ]
} );

webpackConfig.resolveUrlLoader =  {
  root: "./src",
};

webpackConfig.sassLoader = {
  includePaths: project.paths.client( "styles" )
};

webpackConfig.postcss = [
  cssnano( {
    autoprefixer: {
      add: true,
      remove: true,
      browsers: [ "last 2 versions" ]
    },
    discardComments: {
      removeAll: true
    },
    discardUnused: false,
    mergeIdents: false,
    reduceIdents: false,
    safe: true,
    sourcemap: true
  } )
];

// File loaders
/* eslint-disable */
webpackConfig.module.loaders.push(
  { test: /\.woff(\?.*)?$/,  loader: 'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff' },
  { test: /\.woff2(\?.*)?$/, loader: 'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff2' },
  { test: /\.otf(\?.*)?$/,   loader: 'file?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=font/opentype' },
  { test: /\.ttf(\?.*)?$/,   loader: 'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/octet-stream' },
  { test: /\.eot(\?.*)?$/,   loader: 'file?prefix=fonts/&name=[path][name].[ext]' },
  { test: /\.svg(\?.*)?$/,   loader: 'url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=image/svg+xml' },
  { test: /\.(png|jpg)$/,    loader: 'url?limit=8192' }
)
/* eslint-enable */

// ------------------------------------
// Finalize Configuration
// ------------------------------------
// when we don't know the public path (we know it only when HMR is enabled [in development]) we
// need to use the extractTextPlugin to fix this issue:
// http://stackoverflow.com/questions/34133808/webpack-ots-parsing-error-loading-fonts/34133809#34133809
if( !__DEV__ ) {
  debug( "Applying ExtractTextPlugin to CSS loaders." );
  webpackConfig.module.loaders.filter( ( loader ) =>
    loader.loaders && loader.loaders.find( ( name ) => /css/.test( name.split( "?" )[0] ) )
  ).forEach( ( loader ) => {
    const first = loader.loaders[0];
    const rest = loader.loaders.slice( 1 );
    loader.loader = ExtractTextPlugin.extract( first, rest.join( "!" ) );
    delete loader.loaders;
  } );

  webpackConfig.plugins.push(
    new ExtractTextPlugin( "[name].[contenthash].css", {
      allChunks: true
    } )
  );
}

module.exports = webpackConfig;
```

**What is the expected behavior?**
should migrate the file

**Please mention other relevant information such as the browser version, Node.js version, Operating System and programming language.**
`npm version`
```
{ npm: '3.10.10',
  ares: '1.10.1-DEV',
  http_parser: '2.7.0',
  icu: '58.2',
  modules: '48',
  node: '6.10.3',
  openssl: '1.0.2k',
  uv: '1.9.1',
  v8: '5.1.281.101',
  zlib: '1.2.11' }
```
windows 10 1703

**Comment by pksjce** [comment]
@zowers  - Thanks for trying our migrate tool! I'm sorry you faced this issue. I will look into it ASAP.

**Comment by pksjce** [comment]
It seems our test case is not working well to cover this scenario. @zowers - Your webpack config provides us with a great e2e test. I'm planning to give a try to convert this reliably to webpack 2. But would require lot of feedback from you. Will open a PR soon.

**Comment by zowers** [comment]
btw: the config is a product of `react-redux-starter-kit`, also used by the project itself, see https://github.com/davezuko/react-redux-starter-kit/blob/master/config/webpack.config.js

**Comment by zowers** [comment]
seems to be fixed, works now in `webpack-cli` version 1.3.3
